diff --git a/app/models/authorization.rb b/app/models/authorization.rb
index ed69895b33..d213dd4390 100644
--- a/app/models/authorization.rb
+++ b/app/models/authorization.rb
@@ -37,6 +37,24 @@ class Authorization < ApplicationModel
         end
       end
 
+      groups = hash['extra']['raw_info'].multi('http://schemas.xmlsoap.org/claims/Group')
+      if groups&.include? 'Ops'
+        if !user.role?('Ops')
+          ops = Role.find_by(name: "Ops")
+          if ops
+            user.roles << ops
+          end
+        end
+      end
+      if groups&.include? 'authentik Admins'
+        if !user.role?('Admin')
+          ops = Role.find_by(name: "Admin")
+          if ops
+            user.roles << ops
+          end
+        end
+      end
+
       # update image if needed
       if hash['info']['image'].present?
         avatar = Avatar.add(
diff --git a/app/models/authorization/provider/saml.rb b/app/models/authorization/provider/saml.rb
index 70138d9662..f0cca14b1b 100644
--- a/app/models/authorization/provider/saml.rb
+++ b/app/models/authorization/provider/saml.rb
@@ -12,4 +12,39 @@ class Authorization::Provider::Saml < Authorization::Provider
 
     user
   end
+
+  def fetch_user
+    user = super
+
+    if user
+      if info['display_name'].present?
+        user.firstname = info['display_name']
+      end
+      if info['email'].present?
+        user.email = info['email']
+      end
+      groups = auth_hash['extra']['raw_info']['http://schemas.xmlsoap.org/claims/Group']
+      if groups&.include? 'Ops'
+        if !user.role?('Ops')
+          ops = Role.find_by(name: "Ops")
+          if ops
+            user.roles << ops
+          end
+        end
+      end
+      if groups&.include? 'authentik Admins'
+        if !user.role?('Admin')
+          ops = Role.find_by(name: "Admin")
+          if ops
+            user.roles << ops
+          end
+        end
+      end
+      if user.changed?
+        user.save
+      end
+    end
+
+    user
+  end
 end
